File: src/main/java/org/c3lang/intellij/psi/impl/C3StructMemberDeclarationMixinImpl.java
package org.c3lang.intellij.psi.impl;

import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.impl.source.tree.LeafPsiElement;
import com.intellij.psi.stubs.IStubElementType;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.tree.TokenSet;
import com.intellij.psi.util.PsiTreeUtil;
import org.c3lang.intellij.psi.*;
import org.c3lang.intellij.stubs.C3StructMemberDeclarationStub;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public abstract class C3StructMemberDeclarationMixinImpl
	extends C3StubBasedPsiElementBase<C3StructMemberDeclarationStub>
	implements C3StructMemberDeclaration
{
	public C3StructMemberDeclarationMixinImpl(@NotNull ASTNode node)
	{
		super(node);
	}

	public C3StructMemberDeclarationMixinImpl(
		@NotNull C3StructMemberDeclarationStub stub,
		@NotNull IStubElementType<?, ?> nodeType)
	{
		super(stub, nodeType);
	}

	public C3StructMemberDeclarationMixinImpl(
		@NotNull C3StructMemberDeclarationStub stub,
		@Nullable IElementType nodeType,
		@Nullable ASTNode node)
	{
		super(stub, nodeType, node);
	}

	@Override
	public @Nullable PsiElement setName(@NotNull String name)
	{
		LeafPsiElement ident = getNameIdentElement();
		if (ident != null) ident.replaceWithText(name);
		return this;
	}

	@Override
	public @Nullable PsiElement getNameIdentifier()
	{
		return getNameIdentElement();
	}

	@Override
	public @Nullable String getName()
	{
		return getNameIdent();
	}

	@Override
	public int getTextOffset()
	{
		LeafPsiElement ident = getNameIdentElement();
		return ident != null ? ident.getTextOffset() : super.getTextOffset();
	}

	@Override
	public @Nullable LeafPsiElement getNameIdentElement()
	{
		if (getStructBody() != null)
		{
			ASTNode identNode = getNode().findChildByType(C3Types.IDENT);
			return identNode != null ? (LeafPsiElement) identNode.getPsi() : null;
		}
		else
		{
			C3IdentifierList identList = getIdentifierList();
			PsiElement first = identList != null ? identList.getFirstChild() : null;
			return first instanceof LeafPsiElement ? (LeafPsiElement) first : null;
		}
	}

	@Override
	public @Nullable String getNameIdent()
	{
		LeafPsiElement ident = getNameIdentElement();
		return ident != null ? ident.getText() : null;
	}

	@Override
	public @Nullable FullyQualifiedName getDeclaredIn()
	{
		C3DeclaredInProvider provider = PsiTreeUtil.getParentOfType(this, C3DeclaredInProvider.class);
		return provider != null ? provider.getDeclaredIn() : null;
	}

	@Override
	public @Nullable String getStructPath()
	{
		C3StructMemberDeclarationStub stub = getGreenStub();
		return stub != null ? stub.getStructPath() : collectStructPath();
	}

	@Override
	public @Nullable FullyQualifiedName getStructType()
	{
		C3StructMemberDeclarationStub stub = getGreenStub();
		return stub != null ? stub.getStructType() : collectStructType();
	}

	@Override
	public @Nullable FullyQualifiedName getStructPathType()
	{
		C3StructMemberDeclarationStub stub = getGreenStub();
		return stub != null ? stub.getStructPathType() : collectStructPathType();
	}

	@Override
	public @Nullable String getDeclaredInPath()
	{
		C3DeclaredInPathProvider parentProvider =
			PsiTreeUtil.getParentOfType(this, C3DeclaredInPathProvider.class);
		C3IdentifierList identList = getIdentifierList();

		List<String> parts = new ArrayList<>();
		if (parentProvider != null && parentProvider.getDeclaredInPath() != null)
		{
			parts.add(parentProvider.getDeclaredInPath());
		}
		if (identList != null)
		{
			parts.add(identList.getText());
		}
		return String.join(".", parts);
	}

	private @Nullable String collectStructPath()
	{
		String name;
		if (getStructBody() != null)
		{
			ASTNode[] children = getNode().getChildren(TokenSet.create(C3Types.IDENT));
			name = children.length > 0 ? children[0].getText() : null;
		}
		else
		{
			C3IdentifierList identList = getIdentifierList();
			name = identList != null ? identList.getText() : null;
		}

		C3StructMemberDeclaration parentMember =
			PsiTreeUtil.getParentOfType(this, C3StructMemberDeclaration.class);
		String parentPath = parentMember != null ? parentMember.getStructPath() : null;

		List<String> parts = new ArrayList<>();
		if (parentPath != null) parts.add(parentPath);
		if (name != null) parts.add(name);
		String result = String.join(".", parts);
		return result.isEmpty() ? null : result;
	}

	private @Nullable FullyQualifiedName collectStructType()
	{
		C3StructDeclaration structDecl =
			PsiTreeUtil.getParentOfType(this, C3StructDeclaration.class);
		if (structDecl == null) return null;
		String structName = structDecl.getTypeName().getText();
		return new FullyQualifiedName(getModuleDefinition().getModuleName(), structName);
	}

	private @Nullable FullyQualifiedName collectStructPathType()
	{
		if (getStructBody() != null)
		{
			ASTNode[] children = getNode().getChildren(TokenSet.create(C3Types.IDENT));
			if (children.length == 0) return null;
			String innerStructName = children[0].getText();
			FullyQualifiedName st = getStructType();
			if (st == null) return null;
			return new FullyQualifiedName(st.getModule(), st.getName() + "." + innerStructName);
		}
		else
		{
			C3Type type = getType();
			if (type == null) return null;
			return FullyQualifiedName.from(type);
		}
	}
}

File: src/main/java/org/c3lang/intellij/psi/impl/C3StructMemberDeclarationMixinImpl.kt
package org.c3lang.intellij.psi.impl

import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import com.intellij.psi.impl.source.tree.LeafPsiElement
import com.intellij.psi.stubs.IStubElementType
import com.intellij.psi.tree.IElementType
import com.intellij.psi.tree.TokenSet
import com.intellij.psi.util.parentOfType
import com.intellij.util.text.nullize
import org.c3lang.intellij.psi.*
import org.c3lang.intellij.stubs.C3StructMemberDeclarationStub

abstract class C3StructMemberDeclarationMixinImpl : C3StubBasedPsiElementBase<C3StructMemberDeclarationStub>,
    C3StructMemberDeclaration {

    constructor(node: ASTNode) : super(node)

    constructor(
        stub: C3StructMemberDeclarationStub, nodeType: IStubElementType<*, *>
    ) : super(stub, nodeType)

    constructor(
        stub: C3StructMemberDeclarationStub, nodeType: IElementType?, node: ASTNode?
    ) : super(stub, nodeType, node)

    override fun setName(name: String): PsiElement {
        nameIdentElement?.replaceWithText(name)
        return this
    }

    override fun getNameIdentifier(): PsiElement? {
        return nameIdentElement
    }

    override fun getName(): String? {
        return nameIdent
    }

    override fun getTextOffset(): Int {
        return nameIdentElement?.textOffset ?: super.getTextOffset()
    }

    override val nameIdentElement: LeafPsiElement?
        get() {
            return if (structBody != null) {
                node.findChildByType(C3Types.IDENT)?.psi as? LeafPsiElement
            } else {
                identifierList?.firstChild as? LeafPsiElement
            }
        }

    override val nameIdent: String?
        get() = nameIdentElement?.text

    override val declaredIn: FullyQualifiedName?
        get() = parentOfType<C3DeclaredInProvider>()?.declaredIn

    override val structPath: String?
        get() = greenStub?.structPath ?: collectStructPath()

    override val structType: FullyQualifiedName?
        get() = greenStub?.structType ?: collectStructType()

    override val structPathType: FullyQualifiedName?
        get() = greenStub?.structPathType ?: collectStructPathType()

    override val declaredInPath: String?
        get() {
            return listOfNotNull(
                parentOfType<C3DeclaredInPathProvider>()?.declaredInPath,
                identifierList?.text
            ).joinToString(".")
        }

    private fun collectStructPath(): String? {
        val name = if (structBody != null) {
            node.getChildren(TokenSet.create(C3Types.IDENT)).firstOrNull()?.text
        } else {
            identifierList?.text
        }
        return listOfNotNull(
            parentOfType<C3StructMemberDeclaration>()?.structPath,
            name
        ).joinToString(".").nullize()
    }

    private fun collectStructType(): FullyQualifiedName? {
        val structName = parentOfType<C3StructDeclaration>()?.typeName?.text ?: return null

        return FullyQualifiedName(
            moduleDefinition.moduleName,
            structName
        )
    }

    private fun collectStructPathType(): FullyQualifiedName? {

        val structPathType = if (structBody != null) {
            val innerStructName = node.getChildren(TokenSet.create(C3Types.IDENT)).firstOrNull()?.text
                ?: return null
            val structType = structType
                ?: return null

            FullyQualifiedName(structType.module, "${structType.name}.$innerStructName")
        } else {
            FullyQualifiedName.from(checkNotNull(type))
        }

        return structPathType
    }
}